EIPS/eip-2666.md

```diff
@@ -0,0 +1,107 @@
+
+## Motivation
+Historic pricing for these functions in EVM does not reflect inner structure of the underlying computations (inner structure of the hash functions).
+
+## Specification
+
+If `block_number >= X` set the following prices to precompiles and Keccak256 opcode:
+  - `SHA256` precompile (address `0x02`) should be priced as `10 + ((len(input) + 8)/64 + 1) * 9`
+  - RIPEMD precompile (address `0x03`) should be priced as `6 + ((len(input) + 8)/64 + 1) * 12`
+  - Keccak256 built-in function should be priced as `13 + (len(input)/136 + 1)*15`
+
+This EIP *ideally* requires that `MODEXP` repricing is [implemented](https://eips.ethereum.org/EIPS/eip-2565) to also accurately reflect that there is implicit compensation for an old `STATICCALL` cost (pre-2046).
+
+### Detailed summary of repricing approach
+
+This EIP relies on two facts:
+- apriori knowledge of the inner strucute of the hash functions
+- benchmarks provided by the client teams for some reasoable range of input lengths for random inputs (random byte strings of a given length)
+
+#### Benchmarks on the most popular clients
+
+Necessary benchmarks for EIP 2666 were provided by the clients and raw form is assembled in [here](https://docs.google.com/spreadsheets/d/1aCQnk7prrp3Mbcf011BE5zZnkbc3Iw7QAixn6mLbKS0/edit?usp=sharing)
+
+- SHA256 precompile
+
+Currently it's `60` gas + `12` gas per `32` byte word (number of words is `ceil(len(input)/word_len)` here and in similar places. If there is no floor or ceil specifier all divisions below are integer divisions (floor divisions)). Proposed formula is `A * ((len(input) + 8) / 64 + 1) + B`, with coefficients below
+
+|   |   | A  | B  |
+|---|---|---|---|
+| Geth  |   | 5  | 3  |
+| OE  |   | 9  | 4  |
+| Besu  |   | 5  | 10  |
+| Nethermind  |   | 10  | 5  |
+
+EIP-2666 proposes `A = 9`, `B = 10`. There are no large one-off costs in this precompile, so it's EIP-2046 - safe.
+
+- RIPEMD precompile
+
+Currently it's `600` gas + `120` gas per `32` byte word. Proposed formula is `A * ((len(input) + 8) / 64 + 1) + B`, with coefficients below
+
+|   |   | A  | B  |
+|---|---|---|---|
+| Geth  |   | 12  | 6  |
+| OE  |   | 8  | 2  |
+| Besu  |   | 29  | 16  |
+| Nethermind  |   | 10  | 6  |
+
+EIP-2666 proposes `A = 12`, `B = 6`. There are no large one-off costs in this precompile, so it's EIP-2046 - safe. Besu expects to have performance improvements by the end of the year.
+
+- Keccak256 performance
+
+Currently it's `30` gas + `6` gas per `32` byte word. Proposed formula is `A * (len(input) / 136 + 1) + B`, with coefficients below
+
+|   |   | A  | B  |
+|---|---|---|---|
+| Geth  |   | 13  | 13  |
+| OE  |   | 15  | 2  |
+| Besu  |   | 19  | 28  |
+| Nethermind  |   | 16  | 3  |
+
+EIP-2666 proposes `A = 15`, `B = 13`. There are no large one-off costs in this precompile, so it's EIP-2046 - safe. Besu expects to have performance improvements by the end of the year.
+
```

Comment:
All of this should go in the Rationale.  The Specification should *only* contain the specification, not the rationale behind *why* particular values were chosen.  This content is all quite appropriate for Rationale section though.